pyomo/contrib/pynumero/interfaces/external_grey_box.py

```diff
@@ -315,6 +323,29 @@ def evaluate_jacobian_outputs(self):
     # def evaluate_hessian_outputs(self):
     #
 
+    # Support for objectives
+    def has_objective(self):
+        return False
+
+    def evaluate_objective(self) -> float:
+        """
+        Compute the objective from the  values set in
+        input_values
+        """
+        raise NotImplementedError(
+            'evaluate_objective called but not implemented in the derived class.'
+        )
+
+    def evaluate_grad_objective(self, out=None):
+        """
+        Compute the gradient of the objective from the
+        values set in input_values
+        """
+        raise NotImplementedError(
+            'evaluate_grad_objective called but not '
+            'implemented in the derived class.'
+        )
+
```

Comment:
Remind me: Why don't we include default methods for `evaluate_hessian*` that raise `NotImplementedError`? Is the presence of such a method used to determine Hessian support at some point?

Reply:
Looks like it: https://github.com/Pyomo/pyomo/blob/e75a1be9b4ee5e6e3b45beb2f27e21bb154ba5ef/pyomo/contrib/pynumero/interfaces/pyomo_grey_box_nlp.py#L578. If we want to change that, I can create an issue.

Reply:
I prefer `try`/`except NotImplementedError` to using `hasattr`, but I'm not sure it's important enough to change.